nav2_util/src/dump_params.cpp

```diff
@@ -0,0 +1,462 @@
+  {
+    throw std::runtime_error(std::string("service call to \"") + node_name + "\" failed");
+  }
+
+  return result_future.get()->result.names;
+}
+
+static std::vector<rcl_interfaces::msg::ParameterValue>
+get_param_values_for_node(
+  rclcpp::Node::SharedPtr node, std::string node_name,
+  std::vector<std::string> & param_names)
+{
+  auto client = node->create_client<rcl_interfaces::srv::GetParameters>(
+    node_name + "/get_parameters");
+
+  while (!client->wait_for_service(std::chrono::seconds(1))) {
+    if (!rclcpp::ok()) {
+      throw std::runtime_error("client interrupted while waiting for service to appear.");
+    }
+
+    throw std::runtime_error(std::string("GetParameters service for ") +
+            node_name + " not available");
+  }
+
+  auto request = std::make_shared<rcl_interfaces::srv::GetParameters::Request>();
+  request->names = param_names;
+
+  auto result_future = client->async_send_request(request);
+
+  if (rclcpp::spin_until_future_complete(node, result_future, 1s) !=
+    rclcpp::executor::FutureReturnCode::SUCCESS)
+  {
+    throw std::runtime_error(std::string("service call to \"") + node_name + "\" failed");
+  }
+
+  return result_future.get()->values;
+}
+
+static std::vector<rcl_interfaces::msg::ParameterDescriptor>
+get_param_descriptors_for_node(
+  rclcpp::Node::SharedPtr node, std::string node_name,
+  std::vector<std::string> & param_names)
+{
+  auto client = node->create_client<rcl_interfaces::srv::DescribeParameters>(
+    node_name + "/describe_parameters");
+
+  while (!client->wait_for_service(std::chrono::seconds(1))) {
+    if (!rclcpp::ok()) {
+      throw std::runtime_error("client interrupted while waiting for service to appear.");
+    }
+
+    throw std::runtime_error(std::string("DescribeParameters service for ") +
+            node_name + " not available");
+  }
+
+  auto request = std::make_shared<rcl_interfaces::srv::DescribeParameters::Request>();
+  request->names = param_names;
+
+  auto result_future = client->async_send_request(request);
+
+  if (rclcpp::spin_until_future_complete(node, result_future, 1s) !=
+    rclcpp::executor::FutureReturnCode::SUCCESS)
+  {
+    throw std::runtime_error(std::string("service call to \"") + node_name + "\" failed");
+  }
+
+  return result_future.get()->descriptors;
+}
+
+// A local version to avoid trailing zeros
+template<typename T>
+static std::string to_string(const T a_value)
+{
+  std::ostringstream out;
+  out << a_value;
+  return out.str();
+}
+
+static void
+print_yaml(
+  std::string node_name, std::vector<std::string> & param_names,
+  std::vector<rcl_interfaces::msg::ParameterValue> & param_values,
+  std::vector<rcl_interfaces::msg::ParameterDescriptor> & param_descriptors, bool verbose)
+{
+  std::cout << node_name << ":" << std::endl;
+  std::cout << "  ros__parameters:" << std::endl;
+
+  for (unsigned i = 0; i < param_names.size(); i++) {
+    std::function<std::string(void)> pf;
```

Comment:
Is for to using a function instead of a stream?

Reply:
Yeah, I wanted to treat the parameter string as a logical unit. I was thinking of two phases - first determine what the parameter output would be (since this varies quite a bit) and then output the whole parameter block, including the parameter string.